### DISCUSSION TITLE
Setting up Firefly III with NPM and Authentik

### DISCUSSION THREAD

**Discussion by mgrimace** [firefly-iii/firefly-iii#8640] [General]
Here's specifically what I did to get this working in case others run into the barriers that I ran into with this specific setup. Please let me know if there's a better way to do this, or if I'm missing anything (I'm very much a copy/paste novice):

### setup the .env
- Line 38:`TRUSTED_PROXIES=**`, you could probably use the container name `npm` or in my case `npm-attachement` because I also use open-appsec, but `**` works.
- Line 221: `AUTHENTICATION_GUARD=remote_user_guard` 
- Lines 226, 227: 
```
AUTHENTICATION_GUARD_HEADER=HTTP_X_AUTHENTIK_EMAIL
AUTHENTICATION_GUARD_EMAIL=HTTP_X_AUTHENTIK_EMAIL
```
- If you are making these changes later, be sure to restart the containers

### Change ports in the compose.yaml that conflict with NPM
I had to change the ports for `app` (firefly_iii_core) and `importer` (firefly_iii_importer), because I have NPM running already on 80 and 81 already. 
- Under `ports` for the `app`, I changed it to from `'80:8080'` to `'8081:8080'` meaning, my webUI will be accessible at IP:8081
- Under `ports` for the `importer`, I changed it from `'81:8080'` to`'8082:8080'`
- Noting that both 8081 and 8082 aren't use by any of my other containers. 

### Setup Authentik
- Create a provider, name `Firefly ForwardAuth`, Authorization flow `...implicit-consent`, and choose Forward auth (single application). Enter your external URL, e.g., `budget.mydomain.com`
- Create an application, name `Firefly`, slug `firefly`, provider `Firefly ForwardAuth`, UI settings, add an [icon](https://github.com/walkxcode/dashboard-icons/blob/main/png/firefly.png)
- Go to Outposts, and add Firefly to your outpost

### Setup NPM
- add a Proxy Host, domain name, e.g., `budget.mydomain.com`, I selected `block common exploits` and `websockets support`. 
- Typically, I use the container name for `Forward Hostname / IP`, but I had to use the `IP: Port` maybe because I changed the default ports above (not sure).
- go to SSL, and set this up however you normally do it (e.g., cloudflare), I selected `force SSL` and `HTTP/2 Support`.
- go to Advanced, here you can use the 'usual' authentik custom Nginx Configuration, but we need to add the following headers, under `location`, and under `#translate headers from the outposts back to the actual upstream`. Specifically, I added them after the last group, before the closing `}`:

```yaml
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Server $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $host;
    client_max_body_size 64M;
    proxy_read_timeout 300s;
```
- Save
- Probably restart all your containers just to be safe. 

### Things I did not change, but I'm unsure
In the important installation instructions we have the following:
- Change FIREFLY_III_URL in .importer.env to http://app:8080
- Change VANITY_URL in .importer.env to http://localhost

I kept those exactly as it is shown in the instructions.

in the .env there's the following, which I also did not change: 
```yaml
# Please make sure this URL matches the external URL of your Firefly III installation.
# It is used to validate specific requests and to generate URLs in emails.
#
APP_URL=http://localhost
```

Hopefully that helps others in similar situations, and again, if I'm in error or anything I've done above compromised safety/security please point it out.

**Comment by JC5** [comment]
Very nice, thanks!

**Comment by D3SL** [comment]
Thank you, this is what finally helped me get my setup configured properly to work with caddy-docker-proxy or swag.

**Comment by ypsilonkah** [comment]
Thank you for sharing your config. If you are using a mobile app (Abacus or Waterfly III) that accesses Firefly III via auth-token, you need to add `^/api/.*` to the "unauthenticated paths" settings in the Authentik provider configuration

**Reply by mgrimace** [reply]
Thanks so much, and I was unaware of those apps, I appreciate the addition and the recommendations!

**Reply by zxckoko** [reply]
@ypsilonkah , what's your config look like? I am getting errors when add that route. Pls help.

**Comment by evilprophet** [comment]
Hello @mgrimace 
I tried with your setup, but what is "'usual' authentik custom Nginx Configuration"? I'm not sure how I should configure NPM.

**Reply by mgrimace** [reply]
by 'usual' Authentik custom configuration, what I mean is the code snippet that is provided by Authentik for your provider. For example, let's say you make a FireFly ForwardAuth provider. If you click or select it, under the `overview` tab, scroll down to `setup` and you'll see the advanced config options for nginx, NPM, traefik, etc. Select the one for NPM, and make sure to change the line from
`proxy_pass              http://outpost.company:9000/outpost.goauthentik.io;` to `proxy_pass              http://[YourAuthentikIP]:9000/outpost.goauthentik.io;`

**Reply by mgrimace** [reply]
Correction: what I mean to say is copy the advanced config from the provider entry, then add what I put in my original post as an addition along with your AuthentikIP:Port for the `proxy_pass`. If you've never used Authentik, then you'll have to check out their documentation for how to set that all up. Note, Authentik isn't needed to reverse proxy FireFly, I just like the added layer of security personally.

**Reply by evilprophet** [reply]
Thanks for fast reply. I was not aware of changing IP in this snippet. Will check it tomorrow.
I'm already using Authentik for login to Proxmox, Synology and Mealie but config for these services is different.

**Reply by mgrimace** [reply]
Assuming you're using Authentik as a single application forward auth provider, Mealie will likely be the most similiar. Just take that advanced config, and add the snippet from my original post to its `location` section. I use Mealie with Authentik and AFAIK there were no additional custom settings in the advanced config for its NPM (in my setup)

**Reply by evilprophet** [reply]
:( I think something if wrong with my Nginx Proxy Manager.
I tried to use a snippet from Authentik in "Custom Nginx Configuration" field, of course with replacing values with my proper domain/IP. But always after save page is showed as a Offline.

```
# Increase buffer size for large headers
# This is needed only if you get 'upstream sent too big header while reading response
# header from upstream' error when trying to access an application protected by goauthentik
proxy_buffers 8 16k;
proxy_buffer_size 32k;

location / {
    # Put your proxy_pass to your application here
    proxy_pass          $forward_scheme://$server:$port;

    # authentik-specific config
    auth_request        /outpost.goauthentik.io/auth/nginx;
    error_page          401 = @goauthentik_proxy_signin;
    auth_request_set $auth_cookie $upstream_http_set_cookie;
    add_header Set-Cookie $auth_cookie;

    # translate headers from the outposts back to the actual upstream
    auth_request_set $authentik_username $upstream_http_x_authentik_username;
    auth_request_set $authentik_groups $upstream_http_x_authentik_groups;
    auth_request_set $authentik_email $upstream_http_x_authentik_email;
    auth_request_set $authentik_name $upstream_http_x_authentik_name;
    auth_request_set $authentik_uid $upstream_http_x_authentik_uid;

    proxy_set_header X-authentik-username $authentik_username;
    proxy_set_header X-authentik-groups $authentik_groups;
    proxy_set_header X-authentik-email $authentik_email;
    proxy_set_header X-authentik-name $authentik_name;
    proxy_set_header X-authentik-uid $authentik_uid;
}

# all requests to /outpost.goauthentik.io must be accessible without authentication
location /outpost.goauthentik.io {
    proxy_pass          http://outpost.company:9000/outpost.goauthentik.io;
    # ensure the host of this vserver matches your external URL you've configured
    # in authentik
    proxy_set_header    Host $host;
    proxy_set_header    X-Original-URL $scheme://$http_host$request_uri;
    add_header          Set-Cookie $auth_cookie;
    auth_request_set    $auth_cookie $upstream_http_set_cookie;
}

# Special location for when the /auth endpoint returns a 401,
# redirect to the /start URL which initiates SSO
location @goauthentik_proxy_signin {
    internal;
    add_header Set-Cookie $auth_cookie;
    return 302 /outpost.goauthentik.io/start?rd=$request_uri;
    # For domain level, use the below error_page to redirect to your authentik server with the full redirect path
    # return 302 https://authentik.company/outpost.goauthentik.io/start?rd=$scheme://$http_host$request_uri;
}```

**Reply by ypsilonkah** [reply]
I had several issues with Authentik's provided configs. When you go to the overview in Nginx Proxy Manager you can hover your cursor over the red light and it tells you the errormessage. That helped me to finetune it. In authentik's original code lines there where some directives that threw errors in NPM. After deleting them, it turned green and worked.

However, your config differs only slightly from my working config. I think your config is missing the suggested additions by @mgrimace:
```
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Server $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $host;
    client_max_body_size 64M;
    proxy_read_timeout 300s;
```
Further, just to be sure: What did you put at `proxy_pass`? I had some issues where the FQDN did not work (i.e. `proxy_pass http://auth.example.com`). I had to explicitly use my internal IP and respective port number that I setup in docker (e.g. `proxy_pass http://10.10.10.2:8123`)

**Reply by evilprophet** [reply]
I found why page was marked as offline.
On hower it was a hint that this part caused error:
```# Upgrade WebSocket if requested, otherwise use keepalive
map $http_upgrade $connection_upgrade_keepalive {
    default upgrade;
    ''      '';
}
```

When I removed this and line `proxy_set_header Connection $connection_upgrade_keepalive;` site is online.

I also fixed `proxy_pass` because I forgot that I changed default port 9000 so I used domain.

And Now I have just error 500 on Firefly.
```
location / {
    # Put your proxy_pass to your application here
    proxy_pass          $forward_scheme://$server:$port;
    # Set any other headers your application might need
    # proxy_set_header Host $host;
    # proxy_set_header ...
    # Support for websocket
    proxy_set_header Upgrade $http_upgrade;

    ##############################
    # authentik-specific config
    ##############################
    auth_request     /outpost.goauthentik.io/auth/nginx;
    error_page       401 = @goauthentik_proxy_signin;
    auth_request_set $auth_cookie $upstream_http_set_cookie;
    add_header       Set-Cookie $auth_cookie;

    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Server $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $host;
    client_max_body_size 64M;
    proxy_read_timeout 300s;

    # translate headers from the outposts back to the actual upstream
    auth_request_set $authentik_username $upstream_http_x_authentik_username;
    auth_request_set $authentik_groups $upstream_http_x_authentik_groups;
    auth_request_set $authentik_email $upstream_http_x_authentik_email;
    auth_request_set $authentik_name $upstream_http_x_authentik_name;
    auth_request_set $authentik_uid $upstream_http_x_authentik_uid;

    proxy_set_header X-authentik-username $authentik_username;
    proxy_set_header X-authentik-groups $authentik_groups;
    proxy_set_header X-authentik-email $authentik_email;
    proxy_set_header X-authentik-name $authentik_name;
    proxy_set_header X-authentik-uid $authentik_uid;
}

# all requests to /outpost.goauthentik.io must be accessible without authentication
location /outpost.goauthentik.io {
    proxy_pass              https://authentik.xx.xx/outpost.goauthentik.io;
    # ensure the host of this vserver matches your external URL you've configured
    # in authentik
    proxy_set_header        Host $host;
    proxy_set_header        X-Original-URL $scheme://$http_host$request_uri;
    add_header              Set-Cookie $auth_cookie;
    auth_request_set        $auth_cookie $upstream_http_set_cookie;
    proxy_pass_request_body off;
    proxy_set_header        Content-Length "";
}

# Special location for when the /auth endpoint returns a 401,
# redirect to the /start URL which initiates SSO
location @goauthentik_proxy_signin {
    internal;
    add_header Set-Cookie $auth_cookie;
    return 302 /outpost.goauthentik.io/start?rd=$request_uri;
    # For domain level, use the below error_page to redirect to your authentik server with the full redirect path
    # return 302 https://authentik.xx.xx/outpost.goauthentik.io/start?rd=$scheme://$http_host$request_uri;
}
```

**Reply by mgrimace** [reply]
In my experience, 500 error comes from using the domain vs., the IP:Port in your proxy_pass as @ypsilonkah mentioned. If you're going to use a domain here, there's more you need to do.

Try changing it back to IP:Port, and be sure to use whatever your mapped port. If you're using non-standard ports and/or setup, it may be worth seeking Authentik support through Authentik's github discussions.

**Reply by evilprophet** [reply]
I found something different now. 
https://github.com/goauthentik/authentik/issues/8956
I have the same situation :/

----
~Now I have strange situation.
When I open my domain firefly.xx.xx, I'm redirected to the Authentik, when I login I'm redirected back to Firefly but not to domain but to IP. With error "The guard header was unexpectedly empty. See the logs. "~

Nevermind... I changed config few times and I removed additional config mentioned in the first post. Now everything seems to be fine.
And previous problem with error 500 was caused by wrong `TRUSTED_PROXIES` configured in .env file.

**Reply by mgrimace** [reply]
FYI instead of IP:port you can use container-name:port e.g., `authentik-server:9000`

**Comment by jkano** [comment]
Hi @mgrimace!

Thanks for this post, I was able to configure Firefly with Authentik and NPM. But now I have one question, how to handle logout?

After following all the instructions I can login to firefly using Authentik but when I go to Firefly/profile/logout it goes to firefly.mydomain.com/logout and then I see this:

<img width="983" alt="Screenshot 2024-08-22 at 2 36 30 PM" src="https://github.com/user-attachments/assets/1162a876-3923-43b8-8c0f-0d5bf3dd9927">


I tried editing my NPM Firefly proxy configuration by adding another location like this:

```
location /logout {
    proxy_pass          http://192.168.1.92:9000/if/flow/logout/;
}
```

(In http://192.168.1.92:9000/if/flow/logout/,  logout is the Authentik's default-logout flow but just renamed)

The idea behind this was to forward firefly.mydomain.com/logout to http://192.168.1.92:9000/if/flow/logout/

After doing that I try again logging out from firefly and now I get:

<img width="284" alt="Screenshot 2024-08-22 at 3 27 37 PM" src="https://github.com/user-attachments/assets/e7221199-eac3-4b2e-9c95-822b5dde7ada">

Not sure what else should I add to the nginx configuration so this logout flow works.

**Reply by JC5** [reply]
Could this be the solution?

https://github.com/firefly-iii/firefly-iii/blob/main/.env.example#L249

**Reply by jkano** [reply]
Hi @JC5, how should I use that option?

I added `CUSTOM_LOGOUT_URL=http://auth.mydomain/if/flow/logout` to my **env** file, restarted the Firefly docker container and when I go to Profile > Options > logout it stills make a POST request to `http://firefly.mydomain.com/logout` instead of the configured logout URL. 

![Captura de pantalla 2024-08-23 130925](https://github.com/user-attachments/assets/81f8bd20-ff81-469b-8435-b7e39f7fb0f5)

**Reply by JC5** [reply]
That is what should happen, but I see it doesn't work as expected. I will pick that up as a bug!

**Reply by marcelbrueckner** [reply]
> restarted the Firefly docker container

You need to **re-create** your container. Restarting won't pick up any new env variables.

**Reply by jkano** [reply]
Sorry, when I said restarted I meant by using `docker-compose up -d --force-recreate`.

**Comment by turbcool** [comment]
Followed the guide, but got this error:
```
Auth driver [remote_user_guard] for guard [remote_user_guard] is not defined.
at vendor/laravel/framework/src/Illuminate/Auth/AuthManager.php:99
   95▕         if (method_exists($this, $driverMethod)) {
   96▕             return $this->{$driverMethod}($name, $config);
   97▕         }
   98▕ 
➜99▕         throw new InvalidArgumentException(
  100▕             "Auth driver [{$config['driver']}] for guard [{$name}] is not defined."
  101▕         );
  102▕     }
  103▕
```
I use caddy + authentik. Need help, thanks in advance!

EDIT: I have generated `APP_KEY` incorrectly. Example of a working key: `base64:eGkxY3A0NnB2MXBkOTdsOG1ja3MycnJ1MWY1NGgwaWU=` (generated [here](https://generate-random.org/laravel-key-generator)). Old incorrect key: `u/nxiIXGPEee0Z445MEYs9P2Br8/ykF1IyuBikRi6GoUO4WQzAMfKghb8lpbdkn5`.

**Reply by JC5** [reply]
Weird! Which version of Firefly III are you running?

**Reply by adrillon** [reply]
I can reproduce this issue with 6.1.24. This used to work on 6.1.21.

**Reply by JC5** [reply]
What can you reproduce, exactly? The `remote_user_guard` authentication guard works fine in my test setup.

**Reply by adrillon** [reply]
I managed to identity and fix the issue, I believe this is actually an issue with how Firefly is packaged on my system. I'm using NixOS, which packages things very differently from other distributions. I created a bug report on NixOS' repo ([#362597](https://github.com/NixOS/nixpkgs/issues/362597)).

@turbcool run `artisan clear-compiled`. If you get an error about a missing application encryption key, add `APP_KEY=<whatever your key is>` or `APP_KEY_FILE=<path>` before the command.

For reference, here's how to fix it on NixOS :

1. Identity where in the Nix store FF3 is : `ls /nix/store/ | grep firefly-iii-6.1.24`, on my system this yielded `sgd0j13g5ln60dvnj373yijrwsvzs6ny-firefly-iii-6.1.24` which I assume is the same for everyone.
2. `cd /nix/store/sgd0j13g5ln60dvnj373yijrwsvzs6ny-firefly-iii-6.1.24`
3. `export APP_KEY_FILE=<path to your key file>`
4. `./artisan clear-compiled`

**Reply by turbcool** [reply]
> I managed to identity and fix the issue, I believe this is actually an issue with how Firefly is packaged on my system. I'm using NixOS, which packages things very differently from other distributions. I created a bug report on NixOS' repo ([#362597](https://github.com/NixOS/nixpkgs/issues/362597)).
> 
> @turbcool run `artisan clear-compiled`. If you get an error about a missing application encryption key, add `APP_KEY=<whatever your key is>` or `APP_KEY_FILE=<path>` before the command.
> 
> For reference, here's how to fix it on NixOS :
> 
> 1. Identity where in the Nix store FF3 is : `ls /nix/store/ | grep firefly-iii-6.1.24`, on my system this yielded `sgd0j13g5ln60dvnj373yijrwsvzs6ny-firefly-iii-6.1.24` which I assume is the same for everyone.
> 2. `cd /nix/store/sgd0j13g5ln60dvnj373yijrwsvzs6ny-firefly-iii-6.1.24`
> 3. `export APP_KEY_FILE=<path to your key file>`
> 4. `./artisan clear-compiled`

Thanks for advice! But I couldn't resolve the probem. I use Dockerized version of Firefly (tag - latest), no `atrisan` commands work inside the container, throwing the same message.

**Reply by turbcool** [reply]
> I can reproduce this issue with 6.1.24. This used to work on 6.1.21.

Setting `version-6.1.21` tag for the container did not help.

**Reply by turbcool** [reply]
Seems i figured it out! I have generated `APP_KEY` incorrectly. My old not working key: `u/nxiIXGPEee0Z445MEYs9P2Br8/ykF1IyuBikRi6GoUO4WQzAMfKghb8lpbdkn5`.
Example of a working key: `base64:eGkxY3A0NnB2MXBkOTdsOG1ja3MycnJ1MWY1NGgwaWU=` (generated [here](https://generate-random.org/laravel-key-generator)). Thanks everyone.

**Comment by SenpaiSimon** [comment]
I would recommend this video here to get started: [here](https://www.youtube.com/watch?v=Nh1qiqCYDt4)
And i also set:
`AUTHENTICATION_GUARD_HEADER=HTTP_X_AUTHENTIK_EMAIL`  
`AUTHENTICATION_GUARD_EMAIL=HTTP_X_AUTHENTIK_EMAIL`

instead of the ones listed here, it worked afterwards

**Comment by molnarti** [comment]
Anybody got this working in a multi user scenario?

In single user, authentik with firefly works great. Now i want to add a new user.
 I have enabled User registration in the Firefly settings. But the /register page is not available with external authentication enabled. So have created an additional user via the POST /users endpoint with the same email address that already exists in my Authentik user group. 
However, everytime i try to log in with this second user i get a permission denied message from authentik and looks like firefly is not even involved in the flow. In authentik i have granted all the permissions to the user and also it passes the "Check application access" flow
